File: pkg/runtime/runtimes/ts/pulserpc/tests/test_validation.ts
import { strict as assert } from "assert";
import {
  validateString,
  validateInt,
  validateFloat,
  validateBool,
  validateArray,
  validateMap,
  validateEnum,
  validateStruct,
  validateType,
} from "../validation.js";
import { StructMap, EnumMap } from "../types.js";

function testValidateStringSuccess() {
  assert.deepEqual(validateString("hello"), []);
  assert.deepEqual(validateString(""), []);
  console.log("✓ testValidateStringSuccess");
}

function testValidateStringFailure() {
  const errs = validateString(123);
  assert.equal(errs.length, 1);
  assert.match(errs[0].message, /Expected string/);
  assert.equal(errs[0].path, "");

  const errs2 = validateString(null);
  assert.equal(errs2.length, 1);
  assert.match(errs2[0].message, /Expected string/);
  console.log("✓ testValidateStringFailure");
}

function testValidateIntSuccess() {
  assert.deepEqual(validateInt(0), []);
  assert.deepEqual(validateInt(42), []);
  assert.deepEqual(validateInt(-100), []);
  assert.deepEqual(validateInt(5.0), []);
  console.log("✓ testValidateIntSuccess");
}

function testValidateIntFailure() {
  const errs = validateInt("123");
  assert.equal(errs.length, 1);
  assert.match(errs[0].message, /Expected number for int/);

  const errs2 = validateInt(3.14);
  assert.equal(errs2.length, 1);
  assert.match(errs2[0].message, /Expected integer/);

  assert.equal(validateInt(5.1).length, 1);
  console.log("✓ testValidateIntFailure");
}

function testValidateFloatSuccess() {
  assert.deepEqual(validateFloat(3.14), []);
  assert.deepEqual(validateFloat(42), []);
  assert.deepEqual(validateFloat(-1.5), []);
  console.log("✓ testValidateFloatSuccess");
}

function testValidateFloatFailure() {
  const errs = validateFloat("3.14");
  assert.equal(errs.length, 1);
  assert.match(errs[0].message, /Expected number for float/);

  assert.equal(validateFloat(null).length, 1);
  console.log("✓ testValidateFloatFailure");
}

function testValidateBoolSuccess() {
  assert.deepEqual(validateBool(true), []);
  assert.deepEqual(validateBool(false), []);
  console.log("✓ testValidateBoolSuccess");
}

function testValidateBoolFailure() {
  assert.equal(validateBool(1).length, 1);
  assert.equal(validateBool("true").length, 1);
  console.log("✓ testValidateBoolFailure");
}

function testValidateArraySuccess() {
  const elementValidator = (v: any, p: string) => validateString(v, p);
  assert.deepEqual(validateArray(["a", "b", "c"], elementValidator), []);
  assert.deepEqual(validateArray([], elementValidator), []);
  console.log("✓ testValidateArraySuccess");
}

function testValidateArrayWrongType() {
  const elementValidator = (v: any, p: string) => validateString(v, p);
  const errs = validateArray("not a list", elementValidator);
  assert.equal(errs.length, 1);
  assert.match(errs[0].message, /Expected array/);
  console.log("✓ testValidateArrayWrongType");
}

function testValidateArrayElementValidationFails() {
  const elementValidator = (v: any, p: string) => validateString(v, p);
  const errs = validateArray(["a", 123, "c"], elementValidator);
  assert.equal(errs.length, 1);
  assert.match(errs[0].path, /\[1\]/);
  assert.match(errs[0].message, /Expected string/);
  console.log("✓ testValidateArrayElementValidationFails");
}

function testValidateArrayPathTracking() {
  const elementValidator = (v: any, p: string) => validateString(v, p);
  const errs = validateArray(["a", "b", 42], elementValidator, ".items");
  assert.equal(errs.length, 1);
  assert.equal(errs[0].path, ".items[2]");
  console.log("✓ testValidateArrayPathTracking");
}

function testValidateMapSuccess() {
  const valueValidator = (v: any, p: string) => validateInt(v, p);
  assert.deepEqual(validateMap({ a: 1, b: 2 }, valueValidator), []);
  assert.deepEqual(validateMap({}, valueValidator), []);
  console.log("✓ testValidateMapSuccess");
}

function testValidateMapWrongType() {
  const valueValidator = (v: any, p: string) => validateInt(v, p);
  const errs1 = validateMap("not a dict", valueValidator);
  assert.equal(errs1.length, 1);
  assert.match(errs1[0].message, /Expected object for map/);

  const errs2 = validateMap([], valueValidator);
  assert.equal(errs2.length, 1);
  console.log("✓ testValidateMapWrongType");
}

function testValidateMapValueValidationFails() {
  const valueValidator = (v: any, p: string) => validateInt(v, p);
  const errs = validateMap({ a: "not an int" }, valueValidator);
  assert.equal(errs.length, 1);
  assert.match(errs[0].path, /\[a\]/);
  console.log("✓ testValidateMapValueValidationFails");
}

function testValidateEnumSuccess() {
  assert.deepEqual(validateEnum("kindle", "Platform", ["kindle", "nook"]), []);
  assert.deepEqual(validateEnum("nook", "Platform", ["kindle", "nook"]), []);
  console.log("✓ testValidateEnumSuccess");
}

function testValidateEnumWrongType() {
  const errs = validateEnum(123, "Platform", ["kindle", "nook"]);
  assert.equal(errs.length, 1);
  assert.match(errs[0].message, /Expected string for enum/);
  console.log("✓ testValidateEnumWrongType");
}

function testValidateEnumInvalidValue() {
  const errs = validateEnum("invalid", "Platform", ["kindle", "nook"]);
  assert.equal(errs.length, 1);
  assert.match(errs[0].message, /Invalid value for enum/);
  console.log("✓ testValidateEnumInvalidValue");
}

function testValidateStructSuccess() {
  const allStructs: StructMap = {
    User: {
      fields: [
        { name: "id", type: { builtIn: "string" }, optional: false },
        { name: "name", type: { builtIn: "string" }, optional: false },
      ],
    },
  };
  const allEnums: EnumMap = {};

  const errs = validateStruct(
    { id: "123", name: "Alice" },
    "User",
    allStructs,
    allEnums
  );
  assert.deepEqual(errs, []);
  console.log("✓ testValidateStructSuccess");
}

function testValidateStructMissingRequiredField() {
  const allStructs: StructMap = {
    User: {
      fields: [{ name: "id", type: { builtIn: "string" }, optional: false }],
    },
  };
  const allEnums: EnumMap = {};

  const errs = validateStruct({}, "User", allStructs, allEnums);
  assert.equal(errs.length, 1);
  assert.match(errs[0].message, /Missing required field/);
  assert.equal(errs[0].path, ".id");
  console.log("✓ testValidateStructMissingRequiredField");
}

function testValidateStructOptionalField() {
  const allStructs: StructMap = {
    User: {
      fields: [
        { name: "id", type: { builtIn: "string" }, optional: false },
        { name: "email", type: { builtIn: "string" }, optional: true },
      ],
    },
  };
  const allEnums: EnumMap = {};

  assert.deepEqual(validateStruct({ id: "123" }, "User", allStructs, allEnums), []);
  assert.deepEqual(validateStruct({ id: "123", email: "alice@example.com" }, "User", allStructs, allEnums), []);
  console.log("✓ testValidateStructOptionalField");
}

function testValidateStructWithExtends() {
  const allStructs: StructMap = {
    Base: {
      fields: [{ name: "id", type: { builtIn: "string" }, optional: false }],
    },
    User: {
      extends: "Base",
      fields: [{ name: "name", type: { builtIn: "string" }, optional: false }],
    },
  };
  const allEnums: EnumMap = {};

  assert.deepEqual(
    validateStruct({ id: "123", name: "Alice" }, "User", allStructs, allEnums),
    []
  );

  const errs = validateStruct({ name: "Alice" }, "User", allStructs, allEnums);
  assert.equal(errs.length, 1);
  assert.match(errs[0].message, /Missing required field/);
  assert.equal(errs[0].path, ".id");
  console.log("✓ testValidateStructWithExtends");
}

function testValidateStructCollectsMultipleErrors() {
  const allStructs: StructMap = {
    Person: {
      fields: [
        { name: "username", type: { builtIn: "string" } },
        { name: "age", type: { builtIn: "int" } },
        { name: "email", type: { builtIn: "string" } },
      ],
    },
  };
  const allEnums: EnumMap = {};

  const errs = validateStruct(
    { username: "alice", age: "not-a-number", email: 42 },
    "Person",
    allStructs,
    allEnums
  );
  assert.equal(errs.length, 2);
  assert.equal(errs[0].path, ".age");
  assert.equal(errs[1].path, ".email");
  console.log("✓ testValidateStructCollectsMultipleErrors");
}

function testValidateTypeString() {
  const allStructs: StructMap = {};
  const allEnums: EnumMap = {};
  assert.deepEqual(validateType("hello", { builtIn: "string" }, allStructs, allEnums), []);
  console.log("✓ testValidateTypeString");
}

function testValidateTypeOptionalNone() {
  const allStructs: StructMap = {};
  const allEnums: EnumMap = {};

  assert.deepEqual(validateType(null, { builtIn: "string" }, allStructs, allEnums, true), []);

  const errs1 = validateType(undefined, { builtIn: "string" }, allStructs, allEnums, true);
  assert.equal(errs1.length, 1);
  assert.match(errs1[0].message, /cannot be undefined/);

  const errs2 = validateType(null, { builtIn: "string" }, allStructs, allEnums, false);
  assert.equal(errs2.length, 1);
  assert.match(errs2[0].message, /cannot be null/);
  console.log("✓ testValidateTypeOptionalNone");
}

function testValidateTypeArray() {
  const allStructs: StructMap = {};
  const allEnums: EnumMap = {};
  const typeDef = { array: { builtIn: "string" } };

  assert.deepEqual(validateType(["a", "b"], typeDef, allStructs, allEnums), []);

  const errs = validateType(["a", 123], typeDef, allStructs, allEnums);
  assert.equal(errs.length, 1);
  assert.match(errs[0].path, /\[1\]/);
  console.log("✓ testValidateTypeArray");
}

function testValidateTypeMap() {
  const allStructs: StructMap = {};
  const allEnums: EnumMap = {};
  const typeDef = { mapValue: { builtIn: "int" } };

  assert.deepEqual(validateType({ a: 1, b: 2 }, typeDef, allStructs, allEnums), []);

  const errs = validateType({ a: "not int" }, typeDef, allStructs, allEnums);
  assert.equal(errs.length, 1);
  assert.match(errs[0].path, /\[a\]/);
  console.log("✓ testValidateTypeMap");
}

function testValidateNestedStructInArray() {
  const allStructs: StructMap = {
    Child: {
      fields: [
        { name: "name", type: { builtIn: "string" } },
        { name: "age", type: { builtIn: "int" } },
      ],
    },
    Person: {
      fields: [
        { name: "name", type: { builtIn: "string" } },
        { name: "children", type: { array: { userDefined: "Child" } } },
      ],
    },
  };
  const allEnums: EnumMap = {};

  assert.deepEqual(
    validateType(
      { name: "Alice", children: [{ name: "Bob", age: 10 }, { name: "Charlie", age: 12 }] },
      { userDefined: "Person" },
      allStructs,
      allEnums
    ),
    []
  );

  const errs = validateType(
    { name: "Alice", children: [{ name: "Bob", age: 10 }, { name: "Charlie", age: "twelve" }] },
    { userDefined: "Person" },
    allStructs,
    allEnums
  );
  assert.equal(errs.length, 1);
  assert.equal(errs[0].path, ".children[1].age");
  console.log("✓ testValidateNestedStructInArray");
}

testValidateStringSuccess();
testValidateStringFailure();
testValidateIntSuccess();
testValidateIntFailure();
testValidateFloatSuccess();
testValidateFloatFailure();
testValidateBoolSuccess();
testValidateBoolFailure();
testValidateArraySuccess();
testValidateArrayWrongType();
testValidateArrayElementValidationFails();
testValidateArrayPathTracking();
testValidateMapSuccess();
testValidateMapWrongType();
testValidateMapValueValidationFails();
testValidateEnumSuccess();
testValidateEnumWrongType();
testValidateEnumInvalidValue();
testValidateStructSuccess();
testValidateStructMissingRequiredField();
testValidateStructOptionalField();
testValidateStructWithExtends();
testValidateStructCollectsMultipleErrors();
testValidateTypeString();
testValidateTypeOptionalNone();
testValidateTypeArray();
testValidateTypeMap();
testValidateNestedStructInArray();
console.log("\nAll validation tests passed!");

File: pkg/runtime/runtimes/python/pulserpc/tests/test_validation.py
"""Tests for validation functions"""

import unittest
from pulserpc.validation import (
    validate_string,
    validate_int,
    validate_float,
    validate_bool,
    validate_array,
    validate_map,
    validate_enum,
    validate_struct,
    validate_type,
)
from pulserpc.contract import Contract, build_validation_result
from pulserpc.rpctypes import ValidationError, ValidationResult


class TestValidateString(unittest.TestCase):
    def test_success(self):
        self.assertEqual(validate_string("hello"), [])
        self.assertEqual(validate_string(""), [])

    def test_failure(self):
        errs = validate_string(123)
        self.assertEqual(len(errs), 1)
        self.assertIn("Expected string", errs[0].message)
        self.assertEqual(errs[0].path, "")


class TestValidateInt(unittest.TestCase):
    def test_success(self):
        self.assertEqual(validate_int(0), [])
        self.assertEqual(validate_int(42), [])
        self.assertEqual(validate_int(-100), [])
        self.assertEqual(validate_int(5.0), [])

    def test_failure(self):
        errs = validate_int("123")
        self.assertEqual(len(errs), 1)
        self.assertIn("Expected int", errs[0].message)

        errs2 = validate_int(3.14)
        self.assertEqual(len(errs2), 1)
        self.assertIn("Expected int", errs2[0].message)


class TestValidateFloat(unittest.TestCase):
    def test_success(self):
        self.assertEqual(validate_float(3.14), [])
        self.assertEqual(validate_float(42), [])

    def test_failure(self):
        errs = validate_float("3.14")
        self.assertEqual(len(errs), 1)
        self.assertIn("Expected float", errs[0].message)


class TestValidateBool(unittest.TestCase):
    def test_success(self):
        self.assertEqual(validate_bool(True), [])
        self.assertEqual(validate_bool(False), [])

    def test_failure(self):
        self.assertEqual(len(validate_bool(1)), 1)
        self.assertEqual(len(validate_bool("true")), 1)


class TestValidateArray(unittest.TestCase):
    def test_success(self):
        element_validator = lambda v, p: validate_string(v, p)
        self.assertEqual(validate_array(["a", "b", "c"], element_validator), [])
        self.assertEqual(validate_array([], element_validator), [])

    def test_wrong_type(self):
        element_validator = lambda v, p: validate_string(v, p)
        errs = validate_array("not a list", element_validator)
        self.assertEqual(len(errs), 1)
        self.assertIn("Expected list", errs[0].message)

    def test_element_failure(self):
        element_validator = lambda v, p: validate_string(v, p)
        errs = validate_array(["a", 123, "c"], element_validator)
        self.assertEqual(len(errs), 1)
        self.assertIn("[1]", errs[0].path)

    def test_path_tracking(self):
        element_validator = lambda v, p: validate_string(v, p)
        errs = validate_array(["a", "b", 42], element_validator, ".items")
        self.assertEqual(len(errs), 1)
        self.assertEqual(errs[0].path, ".items[2]")


class TestValidateMap(unittest.TestCase):
    def test_success(self):
        value_validator = lambda v, p: validate_int(v, p)
        self.assertEqual(validate_map({"a": 1, "b": 2}, value_validator), [])
        self.assertEqual(validate_map({}, value_validator), [])

    def test_wrong_type(self):
        value_validator = lambda v, p: validate_int(v, p)
        errs = validate_map("not a dict", value_validator)
        self.assertEqual(len(errs), 1)
        self.assertIn("Expected dict", errs[0].message)

    def test_value_failure(self):
        value_validator = lambda v, p: validate_int(v, p)
        errs = validate_map({"a": "not an int"}, value_validator)
        self.assertEqual(len(errs), 1)
        self.assertIn("[a]", errs[0].path)


class TestValidateEnum(unittest.TestCase):
    def test_success(self):
        self.assertEqual(validate_enum("kindle", "Platform", ["kindle", "nook"]), [])
        self.assertEqual(validate_enum("nook", "Platform", ["kindle", "nook"]), [])

    def test_wrong_type(self):
        errs = validate_enum(123, "Platform", ["kindle", "nook"])
        self.assertEqual(len(errs), 1)
        self.assertIn("Expected string for enum", errs[0].message)

    def test_invalid_value(self):
        errs = validate_enum("invalid", "Platform", ["kindle", "nook"])
        self.assertEqual(len(errs), 1)
        self.assertIn("Invalid value for enum", errs[0].message)


class TestValidateStruct(unittest.TestCase):
    def setUp(self):
        self.structs = {
            "User": {
                "fields": [
                    {"name": "id", "type": {"builtIn": "string"}, "optional": False},
                    {"name": "name", "type": {"builtIn": "string"}, "optional": False},
                ]
            }
        }
        self.enums = {}

    def test_success(self):
        errs = validate_struct(
            {"id": "123", "name": "Alice"},
            "User", self.structs["User"], self.structs, self.enums
        )
        self.assertEqual(errs, [])

    def test_missing_required_fields(self):
        errs = validate_struct(
            {}, "User", self.structs["User"], self.structs, self.enums
        )
        self.assertEqual(len(errs), 2)
        self.assertEqual(errs[0].path, ".id")
        self.assertEqual(errs[1].path, ".name")

    def test_optional_field(self):
        structs = {
            "User": {
                "fields": [
                    {"name": "id", "type": {"builtIn": "string"}, "optional": False},
                    {"name": "email", "type": {"builtIn": "string"}, "optional": True},
                ]
            }
        }
        self.assertEqual(
            validate_struct({"id": "123"}, "User", structs["User"], structs, self.enums),
            []
        )
        self.assertEqual(
            validate_struct({"id": "123", "email": "a@b.com"}, "User", structs["User"], structs, self.enums),
            []
        )

    def test_with_extends(self):
        structs = {
            "Base": {
                "fields": [{"name": "id", "type": {"builtIn": "string"}, "optional": False}],
            },
            "User": {
                "extends": "Base",
                "fields": [{"name": "name", "type": {"builtIn": "string"}, "optional": False}],
            },
        }
        self.assertEqual(
            validate_struct({"id": "123", "name": "Alice"}, "User", structs["User"], structs, self.enums),
            []
        )
        errs = validate_struct({"name": "Alice"}, "User", structs["User"], structs, self.enums)
        self.assertEqual(len(errs), 1)
        self.assertEqual(errs[0].path, ".id")

    def test_collects_multiple_errors(self):
        structs = {
            "Person": {
                "fields": [
                    {"name": "username", "type": {"builtIn": "string"}},
                    {"name": "age", "type": {"builtIn": "int"}},
                    {"name": "email", "type": {"builtIn": "string"}},
                ]
            }
        }
        errs = validate_struct(
            {"username": "alice", "age": "not-a-number", "email": 42},
            "Person", structs["Person"], structs, self.enums
        )
        self.assertEqual(len(errs), 2)
        self.assertEqual(errs[0].path, ".age")
        self.assertEqual(errs[1].path, ".email")


class TestValidateType(unittest.TestCase):
    def test_string(self):
        self.assertEqual(validate_type("hello", {"builtIn": "string"}, {}, {}), [])

    def test_optional_none(self):
        self.assertEqual(validate_type(None, {"builtIn": "string"}, {}, {}, True), [])
        errs = validate_type(None, {"builtIn": "string"}, {}, {}, False)
        self.assertEqual(len(errs), 1)
        self.assertIn("cannot be None", errs[0].message)

    def test_array(self):
        type_def = {"array": {"builtIn": "string"}}
        self.assertEqual(validate_type(["a", "b"], type_def, {}, {}), [])
        errs = validate_type(["a", 123], type_def, {}, {})
        self.assertEqual(len(errs), 1)
        self.assertIn("[1]", errs[0].path)

    def test_map(self):
        type_def = {"mapValue": {"builtIn": "int"}}
        self.assertEqual(validate_type({"a": 1, "b": 2}, type_def, {}, {}), [])
        errs = validate_type({"a": "not int"}, type_def, {}, {})
        self.assertEqual(len(errs), 1)
        self.assertIn("[a]", errs[0].path)

    def test_nested_struct_in_array(self):
        structs = {
            "Child": {
                "fields": [
                    {"name": "name", "type": {"builtIn": "string"}},
                    {"name": "age", "type": {"builtIn": "int"}},
                ]
            },
            "Person": {
                "fields": [
                    {"name": "name", "type": {"builtIn": "string"}},
                    {"name": "children", "type": {"array": {"userDefined": "Child"}}},
                ]
            },
        }
        self.assertEqual(
            validate_type(
                {"name": "Alice", "children": [{"name": "Bob", "age": 10}, {"name": "Charlie", "age": 12}]},
                {"userDefined": "Person"}, structs, {}
            ),
            []
        )
        errs = validate_type(
            {"name": "Alice", "children": [{"name": "Bob", "age": 10}, {"name": "Charlie", "age": "twelve"}]},
            {"userDefined": "Person"}, structs, {}
        )
        self.assertEqual(len(errs), 1)
        self.assertEqual(errs[0].path, ".children[1].age")


class TestBuildValidationResult(unittest.TestCase):
    def test_no_errors(self):
        result = build_validation_result([])
        self.assertTrue(result.valid)
        self.assertIsNone(result.error)
        self.assertIsNone(result.invalid_fields)

    def test_errors_with_paths(self):
        errors = [
            ValidationError(path=".username", message="Missing required field"),
            ValidationError(path=".email", message="Expected string, got int"),
        ]
        result = build_validation_result(errors)
        self.assertFalse(result.valid)
        self.assertIsNotNone(result.error)
        self.assertEqual(result.invalid_fields, [".username", ".email"])

    def test_errors_without_paths(self):
        errors = [
            ValidationError(path="", message="Top-level error"),
        ]
        result = build_validation_result(errors)
        self.assertFalse(result.valid)
        self.assertIsNotNone(result.error)
        self.assertIsNone(result.invalid_fields)


class TestContractValidate(unittest.TestCase):
    def setUp(self):
        idl = {
            "structs": [
                {
                    "name": "Person",
                    "fields": [
                        {"name": "username", "type": {"builtIn": "string"}},
                        {"name": "age", "type": {"builtIn": "int"}},
                        {"name": "email", "type": {"builtIn": "string"}},
                    ],
                },
                {
                    "name": "Status",
                    "fields": [
                        {"name": "code", "type": {"builtIn": "int"}},
                    ],
                },
            ],
            "enums": [
                {
                    "name": "Color",
                    "values": [{"name": "red"}, {"name": "green"}, {"name": "blue"}],
                }
            ],
        }
        self.contract = Contract(idl)

    def test_valid_struct(self):
        result = self.contract.validate("Person", {"username": "alice", "age": 30, "email": "alice@example.com"})
        self.assertTrue(result.valid)

    def test_missing_fields(self):
        result = self.contract.validate("Person", {"username": "alice"})
        self.assertFalse(result.valid)
        self.assertIsNotNone(result.error)
        self.assertIsNotNone(result.invalid_fields)

    def test_invalid_field_type(self):
        result = self.contract.validate("Person", {"username": "alice", "age": "thirty", "email": 42})
        self.assertFalse(result.valid)
        self.assertEqual(len(result.invalid_fields), 2)

    def test_valid_enum(self):
        result = self.contract.validate("Color", "red")
        self.assertTrue(result.valid)

    def test_invalid_enum(self):
        result = self.contract.validate("Color", "yellow")
        self.assertFalse(result.valid)
        self.assertIsNotNone(result.error)

    def test_unknown_type(self):
        result = self.contract.validate("NonExistent", {})
        self.assertFalse(result.valid)


if __name__ == "__main__":
    unittest.main()
